feat: Web Search, Vercel AI SDK & Conversation Model

johannes117 opened the pull request:
This PR introduces major features and architectural improvements, including web search via tool calling, a migration to the Vercel AI SDK, and a data model refactor from "Threads" to "Conversations".

#### Key Changes:

*   **Web Search & Tool Calling:**
    *   Integrated the Vercel AI SDK's tool-calling capabilities to enable web search via the Tavily API.
    *   The AI can now autonomously search the web for up-to-date information when needed.
    *   Added a new `ToolCallDisplay` component to show the live status of tool usage (e.g., "Searching the web for...").
    *   Introduced a UI toggle to enable or disable web search per-chat.

*   **Vercel AI SDK Migration:**
    *   Replaced all custom provider-specific streaming logic (`openai.ts`, `google.ts`, etc.) with the unified `streamText` function from the `ai` package.
    *   This standardizes AI interactions, simplifies backend logic, and provides a robust foundation for handling streaming text and tools.

*   ** "Threads" to "Conversations" Refactor:**
    *   Renamed the `threads` table, functions, and all related variables to the more intuitive `conversations`. This improves clarity throughout the codebase, from the Convex schema to frontend components.

*   ** Architecture Documentation:**
    *   Added a new document (`convex_rules.mdc`) detailing the **"Persistent Reactivity"** pattern. This explains our core architectural choice for handling real-time AI interactions in a resilient, multi-device manner.

![Screenshot 2025-06-16 at 2 38 39 PM](https://github.com/user-attachments/assets/70efe59f-3de3-4647-9b16-bae62a40eded)

johannes117 committed:
feat: Refactor messaging system to support streaming responses and enhance message handling

- Rename `create` mutation to `send` and update its functionality to handle user messages and assistant responses.
- Introduce internal mutations `update` and `finalize` for managing streaming message content and completion status.
- Modify message schema to include `isComplete` field for tracking streaming status.
- Update Chat and Message components to reflect changes in message structure and handle streaming responses.
- Refactor ChatInput to utilize new `send` mutation and manage message sending logic.
- Enhance UI components to support real-time message updates and improve user experience during chat interactions.

johannes117 committed:
feat: Add OpenAI SDK integration and enhance logging in messaging system

- Introduced OpenAI SDK to facilitate AI chat functionality.
- Updated package.json to include the OpenAI dependency.
- Enhanced logging in the messages handler to track request and response flow.
- Refactored message validation to ensure proper structure for AI interactions.
- Improved error handling in the chat action to provide clearer feedback on API issues.
- Added console logs in ChatInput for better debugging during message sending.

johannes117 committed:
feat: Integrate multiple AI SDKs and enhance error handling in chat action

- Replaced OpenAI SDK with support for Google, Anthropic, OpenAI, and OpenRouter SDKs.
- Improved logging to include assistant message ID and API key source.
- Enhanced error handling for missing API keys and unsupported model providers.
- Streamlined message preparation for AI interactions with explicit type casting.
- Updated streaming response handling to utilize the new SDKs for improved performance.

johannes117 committed:
feat: Integrate new AI providers and refactor chat functionality

- Added support for Google, Anthropic, and OpenRouter SDKs alongside existing AI integrations.
- Refactored chat action to utilize new streaming methods for improved message handling.
- Updated package.json to include new dependencies for the added AI providers.
- Enhanced error handling and logging for better debugging and user feedback.
- Removed deprecated OpenAI integration and streamlined message processing logic.

johannes117 committed:
refactor: Update title generation logic and improve message handling

- Adjusted title generation to use a faster Google model and refined instructions for generating titles.
- Trimmed whitespace from generated titles before saving to ensure consistency.
- Enhanced error handling by logging errors without throwing exceptions, allowing chat to continue without a title.
- Updated ChatInput to correctly retrieve model configuration based on the selected model.
- Simplified memoization in the Message component by relying on a deep equal comparison for message updates.

johannes117 committed:
feat: Add query to retrieve threads with last message details

- Introduced a new query `listWithLastMessage` to fetch all threads along with their most recent message.
- Updated `useThreads` hook to utilize the new query for enhanced data retrieval.
- Modified `ChatSidebar` component to display last message information and added a spinner for streaming state indication.
- Updated thread interface in `convex-storage` to include optional last message details.

johannes117 committed:
feat: Add setup documentation for Convex Chat Studio and Persistent Reactivity pattern

- Introduced `CONVEX_SETUP.md` with detailed instructions for deploying the Chat Studio application, including environment variables and database schema.
- Added `PERSISTENT_REACTIVITY.md` to explain the architecture of the Persistent Reactivity pattern used for real-time AI interactions, outlining the message flow and benefits of the approach.
- Enhanced documentation to support developers in understanding the setup and architecture of the application.

johannes117 committed:
feat: Enhance message editing and controls functionality

- Introduced editing capabilities in the Message component, allowing users to switch between view and edit modes.
- Added MessageControls for user messages, enabling copy, edit, and regenerate functionalities based on message role.
- Updated MessageEditor to handle message resubmission and deletion of trailing messages for improved user experience.
- Enhanced memoization in the PreviewMessage component to include message list comparison for better performance.
- Adjusted Messages component to pass the full list of messages for control functionalities.

johannes117 committed:
chore: Rename project from "my-v0-project" to "chatstudio" in package.json

johannes117 committed:
fix: Update message rerun functionality to delete user message

- Adjusted the behavior of the "Rerun" action to treat it like "Edit", now deleting the user message along with subsequent messages.
- Enhanced comments for clarity on the changes made to the message handling logic.

johannes117 committed:
refactor: Clean up logging and error handling in AI and messaging components

- Removed unnecessary console logs and comments from the AI title generation and messaging functions for improved code clarity.
- Enhanced error handling by ensuring that errors are logged without interrupting the chat flow.
- Updated message controls to streamline the deletion process for user and assistant messages during reruns.

johannes117 committed:
style: Adjust Message component layout for improved responsiveness

- Updated the Message component's styling to ensure full-width display for better alignment and user experience.
- Changed the class from a max-width constraint to a full-width setting for message containers.

johannes117 committed:
style: Refine Message component layout for user messages

- Updated the Message component's styling to apply a max-width constraint for user messages while ensuring full-width for other roles.
- Enhanced responsiveness by adjusting class names for better alignment and visual consistency.

johannes117 committed:
feat: Implement web search functionality and enhance message handling

- Added a new web search tool using Tavily API to allow real-time information retrieval during chat interactions.
- Updated message schema to include optional fields for tool calls and outputs, enhancing the messaging system's capabilities.
- Modified AI and messaging components to support tool calls, enabling the assistant to utilize web search results in responses.
- Enhanced the ChatInput and Message components to manage and display tool call information effectively.
- Introduced a toggle for enabling/disabling web search in chat settings for improved user control.

johannes117 committed:
fix: Update dependencies and improve AI tool call handling

- Changed the version of "@tavily/core" in package.json to "latest" for up-to-date features and fixes.
- Enhanced the AI tool call logic in the chat function to better handle messages and tool interactions, ensuring more robust processing of user and assistant messages.
- Refactored the web search tool integration to streamline the search result handling and improve overall response accuracy.
- Adjusted the ChatInput component to modify the button variant for better visual consistency when toggling web search functionality.

johannes117 committed:
fix: Update spelling for Australian English and finalize mutation references

- Changed instances of "finalize" to "finalise" in the messaging logic to adhere to Australian English spelling conventions.
- Updated error messages to use "authorised" instead of "authorized" for consistency.
- Added a note in the defaults.mdc file to specify the use of en-AU spelling.

johannes117 committed:
refactor: Introduce MESSAGE_ROLES constant for improved role management

- Added a new constants file to define MESSAGE_ROLES for user, assistant, system, and data roles.
- Updated all relevant components and functions to utilize MESSAGE_ROLES instead of hardcoded strings, enhancing maintainability and consistency across the codebase.
- Adjusted message validation and handling logic to reflect the new role definitions.

johannes117 committed:
refactor: Introduce MESSAGE_ROLES constant and update message handling

- Added a new constants file to define MESSAGE_ROLES for user, assistant, system, and data roles, improving code readability and maintainability.
- Refactored message role handling across various components and functions to utilize the new MESSAGE_ROLES constant instead of hardcoded strings.
- Updated the message schema and related logic to ensure consistent role usage throughout the application.
- Adjusted spelling from "finalize" to "finalise" for consistency with Australian English.

johannes117 committed:
refactor: Introduce unified tool calling abstraction for AI providers

- Extracted tool calling logic into a new `tools.ts` file, enabling a consistent interface for OpenAI, Anthropic, Google, and OpenRouter.
- Implemented the `ToolCallingProvider` class to manage tool execution and streamline interactions with various AI providers.
- Defined standardized tool and tool call interfaces to enhance type safety and maintainability.
- Updated the `ai.ts` file to utilize the new abstraction, improving the handling of tool calls and responses.
- Added documentation in `tools/README.md` to outline the new structure and usage of the tool calling system.

johannes117 committed:
feat: Enhance tool calling logic for Anthropic and Google AI providers

- Implemented detailed handling of tool calls and results for Anthropic, including follow-up message processing.
- Refactored Google AI tool call logic to support function calls and improved response handling.
- Streamlined the integration of tool calls within the messaging flow, ensuring proper event emission for tool usage and results.
- Updated the final content yielding mechanism to provide a more cohesive response structure.

johannes117 committed:
feat: Implement Persistent Reactivity pattern and enhance AI interaction flow

- Added documentation for the Persistent Reactivity architectural pattern, detailing the message flow and benefits of a database-centric approach for real-time AI interactions.
- Updated the AI system prompt to include current date and web search instructions, improving the assistant's contextual awareness.
- Enhanced the chat function to manage tool calls more effectively, ensuring proper message updates during AI interactions.
- Introduced a web search toggle in the ChatRunSettings component, allowing users to enable or disable web search functionality for real-time information retrieval.

johannes117 committed:
refactor: Streamline AI chat functionality and remove deprecated OpenAI integration

- Refactored the chat function to utilize a new tool calling abstraction, enhancing the handling of tool calls and responses.
- Removed the deprecated OpenAI streaming logic and integrated the OpenAI SDK for improved message processing.
- Updated the API key retrieval logic to support multiple providers, including OpenAI, Anthropic, and Google.
- Enhanced error handling and logging for better debugging during AI interactions.
- Cleaned up the codebase by deleting the obsolete OpenAI provider file.

johannes117 committed:
refactor: Revise AI chat structure and introduce conversation management

- Refactored the AI chat functionality to support multiple providers, including Google and Anthropic, enhancing flexibility in AI interactions.
- Introduced a new `updateTitle` mutation for managing conversation titles, improving the overall conversation management system.
- Updated the message schema to reflect changes in conversation handling, replacing thread references with conversation identifiers.
- Removed obsolete provider files and streamlined the codebase for better maintainability and clarity.
- Enhanced error handling and logging throughout the chat and title generation processes for improved debugging.

johannes117 committed:
refactor: Transition from thread to conversation model and enhance message handling

- Replaced all instances of "thread" with "conversation" across the codebase to align with the new conversation management structure.
- Updated message handling logic to utilize conversation identifiers, improving clarity and consistency in data management.
- Introduced new conversation-related queries and mutations, enhancing the overall functionality of the chat system.
- Removed the obsolete threads file and related logic, streamlining the codebase for better maintainability.
- Enhanced UI components to reflect the transition to conversations, ensuring a seamless user experience.

johannes117 committed:
refactor: Enhance message filtering in chat function and remove obsolete tools file

- Implemented filtering logic to exclude empty assistant placeholder messages from the message history, ensuring compliance with Anthropic's requirements.
- Updated the message preparation process to utilize the filtered message history for improved clarity in AI interactions.
- Deleted the unused tools.ts file, streamlining the codebase and removing deprecated tool definitions.

johannes117 committed:
feat: Enhance chat functionality with conversation ID and tool output handling

- Added `conversationId` to `ChatParams` for improved message context in AI interactions.
- Updated the `chat` function to manage tool calls and outputs more effectively, ensuring proper message updates during streaming responses.
- Introduced a new mutation `finaliseToolMessage` for handling tool message finalization.
- Enhanced the `ToolCallDisplay` component to display tool outputs alongside tool calls, improving user feedback during AI interactions.
- Refactored message update logic to accommodate optional content updates, streamlining the message handling process.

johannes117 committed:
refactor: Streamline chat message handling and enhance tool call management

- Refactored the `chat` function to utilize a structured `parts` array for managing message content, tool calls, and results, improving clarity and maintainability.
- Updated the `update` mutation to accept `parts`, allowing for more flexible message updates.
- Enhanced the `finalise` mutation to retain tool call and output data, ensuring a complete message state upon finalization.
- Improved the `ToolCallDisplay` component to handle tool call arguments as objects, enhancing data handling and rendering.
- Adjusted the `Message` and `Messages` components to better manage the rendering of tool calls and outputs, improving user experience during AI interactions.

johannes117 committed:
feat: Introduce system prompt management and refactor prompt creation

- Added a new `prompts.ts` file to centralize system prompt definitions and logic.
- Refactored the `createSystemPrompt` function to accept model-specific instructions and web search options, enhancing flexibility and contextual awareness.
- Updated the `ai.ts` file to utilize the new `createSystemPrompt` function, improving the integration of system prompts in chat interactions.
- Removed the inline system prompt logic from `ai.ts`, streamlining the codebase and improving maintainability.

johannes117 committed:
refactor: Update model-specific prompts for web search tool usage

- Revised the instructions for the `web_search` tool in `prompts.ts` to emphasize mandatory usage for queries related to current events and recent information.
- Enhanced clarity and urgency in the directives, ensuring that models understand the critical nature of using the web search tool for up-to-date responses.

johannes117 committed:
refactor: Enhance message validation and tool call management in AI and conversation handling

- Introduced a structured validation for message parts in `ai.ts`, improving the handling of text, tool calls, and results.
- Updated the message schema to utilize arrays for tool calls and outputs, enhancing data integrity and flexibility.
- Refactored the `chat` function to streamline message processing and ensure proper separation of tool calls and results.
- Added new queries and mutations in `conversations.ts` for better conversation management, including UUID checks and retrieval by ID.
- Implemented a new `messageSummaries.ts` file to manage message summaries, improving the overall architecture for message handling.
- Enhanced error handling across various components to provide clearer feedback during operations.

johannes117 committed:
feat: Implement title generation for initial conversation exchanges in AI chat

- Added logic to generate a title for the conversation after the first user message and assistant response.
- Integrated scheduling for title generation using the appropriate API key based on the provider.
- Enhanced the chat function to improve conversation context and user experience during initial interactions.

johannes117 committed:
fix: Replace Search icon with Globe in ChatInput and ChatRunSettings components

- Updated the ChatInput and ChatRunSettings components to use the Globe icon instead of the Search icon for web search functionality.
- Ensured consistency in UI elements related to web search across the application.

SamDebono commented on convex/ai.ts:
```diff
@@ -119,44 +142,188 @@ export const chat = internalAction({
+          console.log(`[ai.chat] Generating title for conversation ${conversationId}`);
+          
+          // Get the appropriate API key for title generation
+          let googleApiKey = undefined;
```
Need to gracefully manage googleApiKey = undefined

johannes117 replied:
Applied Graceful fallback to hostkey

johannes117 committed:
Merge branch 'main' into vercel-sdk-reimplementation

johannes117 committed:
refactor: Update Chat component imports for improved functionality

- Replaced the use of `useChatRunSettingsStore` with `useConversationByUuid` to streamline conversation handling in the Chat component.
- Enhanced the overall structure and clarity of the Chat component by refining its dependencies.

johannes117 committed:
fix: Update Google API key handling in chat function

- Implemented fallback logic for Google API key usage, allowing the application to use a host API key if the user does not provide one.
- Enhanced title generation process to ensure the correct API key is utilized based on user input or environment settings.
